src/hugr/rewrite.rs

```diff
@@ -0,0 +1,57 @@
+//! Replace operations on the HUGR.
+
+#[allow(clippy::module_inception)] // TODO: Rename?
+pub mod replace;
+use std::mem;
+
+use crate::Hugr;
+pub use replace::{OpenHugr, Replace, ReplaceError};
+
+/// An operation that can be applied to mutate a Hugr
+pub trait Rewrite<E> {
+    /// If `true`, [self.apply]'s of this rewrite guarantee that they do not mutate the Hugr when they return an Err.
+    /// If `false`, there is no guarantee; the Hugr should be assumed invalid when Err is returned.
+    const UNCHANGED_ON_FAILURE: bool;
+
+    /// Checks whether the rewrite would succeed on the specified Hugr.
+    /// If this call succeeds, [self.apply] should also succeed on the same `h`
+    /// If this calls fails, [self.apply] would fail with the same error.
```

Comment:
An alternative is to return `Result<(), Option<E>>` where returning Err(None) means, no guarantee can be given about `apply`. If a rewrite really has to get partway through in order to complete its own validity checks then said alternative would support that. But I think we should leave as `Result<(), E>` until/unless we find a Rewrite where that's difficulty.

(Compound Rewrites is one case. `verify` in such cases might have to involve cloning the hugr.)

Reply:
I'd say custom enums with `{Recoverable, NonRecoverable}` variants rather than `Option<E>`.
But yeah, let's keep it simple for now.